### PR TITLE
Add Deferred HIR Name resolution

### PR DESCRIPTION
**Pull request by DropDemBits**
This separates the process of resolving names from defs from the process of lowering the AST to HIR.
This allows for greater flexibility in how defs are resolved, required for bringing unqualified imports in along with their associated imports, and also modularizes the resolving process under one module.

This also has the side benefit of not having to undeclared defs, and instead now only leaves unnamed def tracking as the only kind of special def tracking we need to do.

### CODE REVIEW

**Comment by DropDemBits** [comment]
Thought up when trying to implement bringing in an import's unqualified exports:

```rs
// ???: By the time we hit an import, do we know for certain what
// `imported_def` might resolve to?
// => We generate name resolution maps for libraries in topo-order,
//    so dependent libraries should already have name res maps generated.

// ???: How do we handle unqualified exports from different libraries?
// We've already frozen the defs by this point, so we can't add new defs
// => The main issue is that ScopeTracker resolves to defs, not necessarily
//    `Resolve`s. When we do find a matching def, that's when we wrap it into
//    a `Resolve`. We might be able to get away with ScopeTracker instead
//    mapping bindings to `Resolve`s, though the only complication is what
//    happens during redeclaration reporting

// ???: If we do need to perform in-place def-resolution, how do we deal with
// being routed through another library?
// Note that this is only required for figuring out if a def has associated exports

// ???: What if we only explored the defs that a binding might encounter?
// E.g. we essentially record what scopes are pushed & popped, and after what
// span a def/resolution is introduced. Then we'd use the span of a binding
// to figure out what scoping commands to "apply", generating a resolution
// scope.
```

**Comment by DropDemBits** [comment]
Probably need to pull resolver logic into hir_db, or a separate crate linked to hir_db

**Comment by DropDemBits** [comment]
Keeping some answers here:

## By the time we hit an import, do we know for certain what `imported_def` might resolve to?
We generate name resolution maps for libraries in topo-sort order, so dependent libraries should already have name res maps generated.


## How do we handle unqualified exports from different libraries?
We've already frozen the defs by this point, so we can't add new defs

The main issue is that `ScopeTracker` resolves to defs, not necessarily `Resolve`s. When we do find a matching def, that's when we wrap it into a `Resolve`. We might be able to get away with `ScopeTracker` instead mapping bindings to `Resolve`s, though the only complication is what happens during redeclaration reporting


## If we do need to perform in-place def-resolution, how do we deal with being routed through another library?
Needs to be performed in two stages:

- Intra library (while we're still constructing the resolution map)
- Inter library (using frozen resolution maps)

Basically, we use local resolution until we hit an external resolution
Once we do, we defer to the db's resolution query for the given library


## What if we only explored the defs that a binding might encounter?
E.g. we essentially record what scopes are pushed & popped, and after what span a def/resolution is introduced. Then we'd use the span of a binding to figure out what scoping commands to "apply", generating a resolution scope.

This is speculation territory that we don't need to explore, so no worries in ignoring it.

**Comment by DropDemBits** [comment]
Will merge as-is, but name resolution changes again! (it will actually be moved back to before HIR lowering, as originally intended)

bors r+
